[feature] Created register menu group functionality

niteshsinha17 opened the pull request:
Currently I have made a function to register menu groups. Also I am showing registered menu groups in the menu. User can navigate throughout the website very easily using the menu.

**Note:** In this commit designing was not my concern.

![Screenshot_2021-03-17 Add user OpenWISP Admin](https://user-images.githubusercontent.com/42765292/111433458-520ba780-8724-11eb-9fe0-9597b91cea4b.png)

closes #189 
closes #190 
closes #193

nemesisdesign commented on tests/test_project/apps.py:
```diff
@@ -15,9 +15,38 @@ class TestAppConfig(ApiAppConfig):
+    def register_default_menu_groups(self):
+        groups = [
+            {
+                'group_name': "List Group",
```
group_name > name, no need to repeat the word "group", since it's implied

nemesisdesign commented on tests/test_project/apps.py:
```diff
@@ -15,9 +15,38 @@ class TestAppConfig(ApiAppConfig):
+        groups = [
+            {
+                'group_name': "List Group",
+                'group': [
```
this could be `models` or `items` if we want something more generic (we can add support for linking pages that are not models so maybe it makes sense to use something generic)

niteshsinha17 commented:
**Functionality Added**

1. Registration of multiple menu groups.
2. Functionality to add new items in the default menu group. I have divided the menu into two sections i.e custom and default.
3. Now it is possible to add icons to the group.
4. It logs a warning when using old register menu items functionality.
5. Now we can pass preference in which order we want to show menu groups.
6. It also logs errors when we apply different preferences to the same group and when we use the same name for the groups multiple times.
7. We can now register both model link and link of another site.

![Screenshot from 2021-04-25 16-34-33](https://user-images.githubusercontent.com/42765292/115991127-b273eb80-a5e4-11eb-8fe7-66c7ecf4a6eb.png)

nemesisdesign commented on openwisp_utils/admin_theme/context_processor.py:
```diff
@@ -7,14 +9,60 @@
 
 def menu_items(request):
     menu = build_menu(request)
+    if len(menu):
```
should be possible to just do `if menu`

nemesisdesign commented on openwisp_utils/admin_theme/context_processor.py:
```diff
@@ -7,14 +9,60 @@
+                model_class = registry.apps.get_model(app_label, model)
+                model_label = model.lower()
+                uuid = item['name']
+                url = reverse('admin:{}_{}_{}'.format(app_label, model_label, uuid))
```
do not use format, use f-strings

nemesisdesign commented on openwisp_utils/admin_theme/templates/admin/base_site.html:
```diff
@@ -45,6 +45,21 @@ <h1 id="site-name"><a href="{% url 'admin:index' %}">{{ site_header }}</a></h1>
     {% endfor %}
+    {% for group in openwisp_menu_groups %}
+        <div class="group-name">{% if group.icon %}
+            <img src={% static group.icon %} alt="icon" />
```
I took for granted that we'd implement icons with CSS since CSS is more powerful and can give us more control, so the icon attribute should just be a class which is used to specify which icon has to be used.

nemesisdesign reviewed (commented):
Is ordering implemented? Keep in mind we need to be able to specify the position of each link.

Can icons be defined also at link level (not only group level)?

nemesisdesign reviewed (commented):
> Functionality to add new items in the default menu group. I have divided the menu into two sections i.e custom and default.

I do not understand this.

> Now we can pass preference in which order we want to show menu groups.

Please ensure it's possible to specify the order of groups and links within the group.

> We can now register both model link and link of another site.

Nice :+1:

niteshsinha17 commented:
> I do not understand this.

I am dividing menu to sections. One of them is default group that will be provided by Utils itself. Anyone can add more links too it.
If you don't want this then I can remove it.

>I took for granted that we'd implement icons with CSS since CSS is more powerful and can give us more control

Okay I will change this.

>Is ordering implemented? Keep in mind we need to be able to specify the position of each link.
Can icons be defined also at link level (not only group level)?

Yes, we can implement both. I will update it again for this.

atb00ker commented on openwisp_utils/admin_theme/context_processor.py:
```diff
@@ -7,14 +9,60 @@
     menu = build_menu(request)
+    if len(menu):
+        logging.warning(
+            'Register_menu_items is outdated. Plase update to use register_menu_group'
```
`outdated` -> `deprecated`

atb00ker commented on openwisp_utils/utils.py:
```diff
@@ -29,6 +29,48 @@ def register_menu_items(items, name_menu='OPENWISP_DEFAULT_ADMIN_MENU_ITEMS'):
+    for group in menu_groups:
+        if menu.get(group['name'], None) is not None:
+            # menu name already present
+            raise Exception(
```
Can we make it `ArguementException` instead of `Exception`, I think that would more accurately represent the error.

atb00ker commented on openwisp_utils/utils.py:
```diff
@@ -29,6 +29,48 @@ def register_menu_items(items, name_menu='OPENWISP_DEFAULT_ADMIN_MENU_ITEMS'):
+            group_map[group['name']] = group['config']
+        else:
+            # provided same group name multiple times
+            raise Exception('Two menu groups can not have same name')
```
Same as above.

atb00ker commented on openwisp_utils/utils.py:
```diff
@@ -29,6 +29,48 @@ def register_menu_items(items, name_menu='OPENWISP_DEFAULT_ADMIN_MENU_ITEMS'):
+    for group in preference:
+        if group_map.get(group, None) is None:
+            # same group has given multiple preference
+            raise Exception('Same group has provided multiple preferences')
```
Same as above.

atb00ker reviewed (commented):
Quick comments

niteshsinha17 commented:
@nemesisdesign and @atb00ker Thanks for waiting. I have updated it again.

1. Now we can add icons to each link level too.
2. We can also provide position of each link.
3. For groups, I am using a preference list so that if users skip this then we don't run any extra piece of code to achieve ordering in groups.
4. I have added extra checks so that if user provide wrong values then we can throw error to make them easy to understand whats wrong they have done.
5. Using f-strings

![Screenshot from 2021-04-28 18-30-11](https://user-images.githubusercontent.com/42765292/116407872-d5ee9e80-a84f-11eb-8fc4-248563267c2d.png)

atb00ker commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+
+from ..utils import SortedOrderedDict
+
+MENU = SortedOrderedDict()
```
A global variable, is there another way?
Perhaps a singleton?

atb00ker replied:
Global variables are evil. We should try to avoid them when possible and only use them when there is no other way.
I think we can continue for now, this is a lower priority issue.
@nemesisdesign What do you think? Should we log an issue for creating a singleton class fir managing this?

P.S: Global variables are evil because their scope is too big and we can't limit it's access/editing, a singleton can give us some control over it's access! :smile:

niteshsinha17 replied:
Thanks @atb00ker 
Lets do it in a separate PR. For now I am updating this PR so that we can proceed further.

nemesisdesign replied:
how are we currently doing it for the current menu? Or for the dashboard with the pie charts?

@atb00ker I would not look into this as something that is evil or not, if it's useful I would use it, if it's harmful or has downsides that affect us then we should avoid it. A data structure which is accessible from code is not something new in python, which does not have any real private methods.

If we want to improve it, we can create a singletone and try to hide the datastructure in the singleton, providing methods to manipulate it, just let's make sure it's consistent with the dashboard code and not end up with having it done in way in one place and another way in another place.

atb00ker replied:
Agreed, Singleton implementation is implementing at lot of places; we can come back to it in the future if we have time. (Nice to have task)

atb00ker commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+class MenuItem(BaseMenuItem):
+    def __init__(self, config):
+        super().__init__(config)
+        if config.get('name', None) is not None:
```
Hey, test if  `if config.get('name', None):` gives the same result, `None` is treated like false!

atb00ker commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+
+    def get_label(self, config=None):
+        if config:
+            if config.get('label', None) is not None:
```
Hey, test if  `if config.get('label', None):` gives the same result, `None` is treated like false!

atb00ker commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+                )
+            self.name = config['name']
+        else:
+            raise ValueError(f'"name" is missing in the config-{config}')
+        if config.get('model', None) is None:
+            raise ValueError(f'"model" is missing in config-{config}')
```
`the` in the first statement not the second! :-)

atb00ker commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+class MenuGroup(BaseMenuItem):
+    def __init__(self, config):
+        super().__init__(config)
+        if config.get('label', None) is None:
```
`if not config.get('label', None):` should work.

nemesisdesign replied:
`if not config.get('label'):` too

atb00ker commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+        if not isinstance(config['label'], str):
+            raise ValueError('"level" should be a type of "str" for config-{config}')
+        self.label = _(config['label'])
+        if config.get('items', None) is None:
```
Sample as above.

atb00ker commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+class MenuLink(BaseMenuItem):
+    def __init__(self, config):
+        super().__init__(config)
+        if config.get('label', None) is None:
```
Same as above.

atb00ker commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+                '"label" is missing for a menu link having config - {config}'
+            )
+        self.label = config['label']
+        if config.get('url', None) is None:
```
Same as above.

atb00ker reviewed (commented):
Partial Review:

niteshsinha17 commented:
@atb00ker 
Thanks, I will correct it soon. Also started designing in a separate branch and will push soon.

atb00ker commented:
That's great! @niteshsinha17

niteshsinha17 commented:
@nemesisdesign and @atb00ker Its ready.
How should we proceed now?

atb00ker reviewed (commented):
Could you please include changes in REAME to explain how can a user register menu group / use the new functionality. :smile:

niteshsinha17 commented:
>Hey, your last comment on this PR is 3 days old, if something is getting stale, please remember to ping someone, it's entirely possible to miss pings on GitHub!

I was going to remind you tomorrow. I thought you were waiting for the results.

>Could you please elaborate the question? Proceed in context of this PR?

I was thinking to start merging them into some branch if things are ok.

>Could you please include changes in REAME to explain how can a user register menu group / use the new functionality.

Sure,  I will add this. I am noting every point so that we can open issues to solve them.

niteshsinha17 commented:
> Could you please include changes in REAME to explain how can a user register menu group / use the new functionality.

@atb00ker Added in README, you can check now.

kkreitmair commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+    def __init__(self, config):
+        super().__init__(config)
+        if config.get('name'):
+            if not isinstance(config['name'], str):
```
I would suggest to use always the `.get` method for accessing dict values through keys, even if there existences is checked previously. This would make the code more resilient to changes later in time.
@nemesisdesign @atb00ker what do you think about it?

nemesisdesign replied:
Agreed, I would do something like:

```python
value = config.get('name')
if not value:
  raise # ...
elif not isinstance(value, str):
   raise # ...
# ...
```

@niteshsinha17 you should not call `config.get('name')` and throw away the value like in that `if`, and then access the dict again later. Try to remember to access the dict once and reuse the value.

niteshsinha17 replied:
Got it. Thanks :+1:

kkreitmair reviewed (commented):
Good work @niteshsinha17 !

As others already mentioned, tests would give this PR a good finish.

nemesisdesign commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+            app_label, model = config['model'].split('.')
+            model_class = registry.apps.get_model(app_label, model)
+            return f'{model_class._meta.verbose_name_plural} {self.name}'
+        return self.label
```
why are you returning `self.label` in the end while returning `config['label']` above?

niteshsinha17 replied:
I think it was a mistake. Fixed it. :+1:

nemesisdesign commented on README.rst:
```diff
@@ -460,6 +460,150 @@ template ``context_processors`` in ``settings.py`` as shown below.
 
+``register_menu_groups``
+^^^^^^^^^^^^^^^^^^^^^^^^
+It allows to register different menu items or groups at different positions
```
leave blank line after heading

niteshsinha17 commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+        ):
+            raise ValueError(
+                'group should be type of "MenuGroup", "MenuItem" or "MenuLink"'
+            )
```
Thanks. Changed the implementation

niteshsinha17 commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,154 @@
+        super().__init__(config)
+        if not config.get('label'):
+            raise ValueError(
+                '"label" is missing in a menu group having config-{config}'
```
Thanks for noticing it.

niteshsinha17 commented:
Added tests. It is ready for testing.

atb00ker commented on README.rst:
```diff
@@ -460,6 +460,154 @@ template ``context_processors`` in ``settings.py`` as shown below.
+        background: url(imageurl);
+    }
+
+**Note**: There are two ways to add links in the navigation menu i.e ``register_menu_groups`` and
```
and we will be deprecating ``register_menu_items`` too, correct? Can we mention that information here to encourage the new functions usage?

atb00ker commented on tests/test_project/tests/test_menu.py:
```diff
@@ -0,0 +1,284 @@
+from openwisp_utils.utils import SortedOrderedDict
+
+
+class TestMenuSchema(TestCase):
```
The tests here seem too big to me, each test is testing a lot of things.
I think having smaller tests is beneficial because
1. they can run in parallel
2. If there a failure in step X another case which is on step X+1 might not be reported. 
3. We can be sure that data is pristine in smaller tests. In larger tests, it's possible to mistakenly taint data.

However, I'd like to know what others thing about it; hence I'd wait for others reviews before recommending this change!

nemesisdesign replied:
I think the size of the test suite is still fine and does not need to be split.
In general, let's try to keep test suites under 1K lines.

> they can run in parallel

Yeah but when not running in parallel, processing more test classes can take more time because the DB is destroyed and recreated each time. So each approach has pros and cons, the right thing to do is to keep it simple do additional steps only when pain makes it really needed.

> If there a failure in step X another case which is on step X+1 might not be reported.

If one subtests fails the others will be executed anyway. 

> We can be sure that data is pristine in smaller tests. In larger tests, it's possible to mistakenly taint data.

Subtests should be used only to perform tests which are related to a specific feature, which I think is what Nitesh is doing here, so I think that's very good.

atb00ker reviewed (commented):
Looks good to me, some comments below:

niteshsinha17 commented on README.rst:
```diff
@@ -460,6 +460,154 @@ template ``context_processors`` in ``settings.py`` as shown below.
+        register_menu_groups,
+    )
+
+    my_group_items = {
```
Yes,  I created it to use but forget to use it. I will fix it.

niteshsinha17 replied:
Reviewing is really very dedicated work. I really appreciate your work 👏

atb00ker replied:
I think this is outdated now, closing.

nemesisdesign reviewed (changes requested):
It's looking good but I think we need to simplify a bit, read on.

niteshsinha17 commented on README.rst:
```diff
@@ -460,6 +460,154 @@ template ``context_processors`` in ``settings.py`` as shown below.
+                    'icon': 'shelf-icon',
+                }
+            )
+        3: MenuLink(config={'label': 'Link Name', 'url': 'https://link.com'})
```
I think its good to have class for each item because it help in writing validations, testing and creation of menu much simpler way.
>We parse the data structure and determine what class to use (if the classes are needed at all), this way users only have to pass a data structure and can ignore these 3 classes. It will result simpler to them. It will also result simpler to document.

But it will help user to debug the errors. In many places we are telling user that they have done some mistake in the config of  a menu item or a menu link or a menu group so they don't need to whole code.They can identify mistake quickly. It also creates a good readability in the code. What do you think about this?

atb00ker replied:
> help in writing validations, testing and creation of menu much simpler way.

I wonder how, can you please elaborate with an example?

> they don't need inspect the whole code

Why would they need to inspect the whole code in using a json? Why don't we let them know the exact item that failed to parse?

niteshsinha17 replied:
>I wonder how, can you please elaborate with an example?

For example, a class object validates itself when they are created. We don't need to parse json to identify the type of menu item. Also, we don't need to parse MENU again before creating its context. we are doing it independently using classes.

>Why would they need to inspect the whole code in using a json? Why don't we let them know the exact item that failed to parse?

We can do this. If you will say then I will change it. Don't you think readability is also important?

niteshsinha17 commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,194 @@
+
+
+class MenuItem(BaseMenuItem):
+    def __init__(self, config):
```
Okay I will add it in each class

niteshsinha17 commented:
@nemesisdesign  and @atb00ker 
I have updated this PR as you have suggested.

atb00ker commented on tests/test_project/apps.py:
```diff
@@ -86,3 +88,60 @@ def register_dashboard_charts(self):
+        auth_group = {
+            'label': 'Authentication And Authorization',
+            'items': auth_list,
+            'icon': 'user-icon',
```
When I was using it, I realised, instead of setting icons like this, shouldn't we use https://icons.getbootstrap.com/ ? Users will have a lot more choices!

niteshsinha17 replied:
@atb00ker Its good idea.
I have checked their [accessibility](https://icons.getbootstrap.com/#accessibility) page.
I think we can use this
```
<i class="bi-alarm" style="font-size: 2rem; color: cornflowerblue;"></i>
```
We will ask user to provide only class name. @nemesisdesign  and @atb00ker What do you think?

atb00ker replied:
Sounds good to me! :smile:

atb00ker replied:
Hey, please create a new issue for it and handle it seperately, otherwise we will be stuck in perpetual review loop.

nemesisdesign replied:
is this resolved?

niteshsinha17 replied:
Created an issue https://github.com/openwisp/openwisp-utils/issues/206 for it.

atb00ker reviewed (commented):
![hard1](https://user-images.githubusercontent.com/24985760/119865702-9348e080-bf39-11eb-895f-e4775c3e5bcb.png)

Two things in the screenshot:
1. Docs section (and code) don't have icons!
2. I think there are 2 things registered in the auth group.
```
        auth_list = {
            2: {
                'label': 'Add user',
                'model': 'auth.User',
                'name': 'changelist',
                'icon': 'add-icon',
            },
            1: {
                'model': 'auth.User',
                'name': 'add',
                'icon': 'edit',
                'label': 'Edit User',
            },
        }
```

I am a superuser, yet I don't see the second one. I am missing something?

niteshsinha17 commented:
> ![hard1](https://user-images.githubusercontent.com/24985760/119865702-9348e080-bf39-11eb-895f-e4775c3e5bcb.png)
> 
> Two things in the screenshot:
> 
> 1. Docs section (and code) don't have icons!
> 2. I think there are 2 things registered in the auth group.
> 
> ```
>         auth_list = {
>             2: {
>                 'label': 'Add user',
>                 'model': 'auth.User',
>                 'name': 'changelist',
>                 'icon': 'add-icon',
>             },
>             1: {
>                 'model': 'auth.User',
>                 'name': 'add',
>                 'icon': 'edit',
>                 'label': 'Edit User',
>             },
>         }
> ```
> 
> I am a superuser, yet I don't see the second one. I am missing something?

I found the issue. There was an indentation problem which was stoping the registration of other item. I will include a test to detect this. Thanks

niteshsinha17 commented:
Update of latest changes:
1. Removed ``register_menu_items`` section from readme.
2. I think ``Model Link`` is a better name. So changed name of ``Menu Item``.
3. Double checked spellings. I hope nothing is left now.
4. I was using ``value error`` to raise exceptions in many places. I think it was not good so changed it to ``ImproperlyConfigured``.
5. Added one more test.
6. Fixed problems in readme.

 Updated task list here https://github.com/openwisp/openwisp-utils/issues/189#issue-894593577

kkreitmair commented on openwisp_utils/admin_theme/templates/admin/base_site.html:
```diff
@@ -45,6 +45,20 @@ <h1 id="site-name"><a href="{% url 'admin:index' %}">{{ site_header }}</a></h1>
+            <div>
+            {% for item in group.sub_items %}
+            <a href="{{ item.url }}"
+               class="ow-menu-link"><i class="bi-alarm"></i><span class="{{item.icon}}"></span>{{ item.label }}</a>
```
I would recommend to use `<i>` tag for presenting the icons. The `<span>` part could be removed in my opinion.

Also rework this changes according indentation and consistent formatting of template variables (for example `{{item.icon}}` should be `{{ item.icon }}`)

niteshsinha17 replied:
Fixed indentation issue.
>Also rework this changes according indentation and consistent formatting of template variables (for example {{item.icon}} should be {{ item.icon }})

We can change it here https://github.com/openwisp/openwisp-utils/pull/185

Thanks @kkreitmair Its ready now but CI is not working

kkreitmair reviewed (commented):
Here some points to fix or to consider from my review.

nemesisdesign commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,16 @@
+- [ ] I have written new test cases to avoid regressions. (if necessary)
+- [ ] I have updated the documentation. (e.g. README.rst)
+
+> #Closes #(issue-number)#
```
this shouldn't be here

atb00ker replied:
@niteshsinha17 
I you should do:
```
git fetch upstream dev
git rebase upstream/dev
git push -f
```
to fix this problem.

niteshsinha17 replied:
I tried but it is still here, I think I should revert it

atb00ker replied:
Are you sure? It should work. 
I think this is a `Select is not broken` situation.
It should work.

I can't try fixing it because this one is open from your fork.
Can you share the output?

niteshsinha17 replied:
Out says, already up to date.

niteshsinha17 replied:
![image](https://user-images.githubusercontent.com/42765292/120823424-0d4e1a80-c575-11eb-8d0e-248ddd085cce.png)

niteshsinha17 replied:
![image](https://user-images.githubusercontent.com/42765292/120823632-3f5f7c80-c575-11eb-951a-26e0c26eb501.png)

nemesisdesign commented on README.rst:
```diff
@@ -460,6 +460,161 @@ template ``context_processors`` in ``settings.py`` as shown below.
+        3: {'label': 'Link Name', 'url': 'https://link.com'},
+    }
+
+    register_menu_groups(my_menu)
```
I would have expected something like we have for the dashboard:
```
register_menu_group(position=1, config=group_config)
``` 

That way we can decide also the position of groups, which is not clear from this syntax.

Did you think about this?

niteshsinha17 replied:
It can be a good option, then user will have call this function again and again. 
@nemesisdesign Would you like to go with this option? 
Also we will have to add ``menu group`` items like we are adding right now. I think current implementation is not having issue. User can understand by reading documentation. But your decision will be final.

nemesisdesign replied:
@niteshsinha17 let's define who's the user. For me users are users using OpenWISP.
This is meant for us, developers, users of this library as developers.
We will not be able to call this function only once, it will have to be called in each app, just like we do now with `register_menu_items`.

How would your solution work in practice?

niteshsinha17 replied:
Yes, you are right. I was thinking about ``model link`` or ``general link``. For this, they will have to call functions multiple times.
Focusing on OpenWISP requirements your solution is good and it is consistent too. So let's stay with your soluton. I will make changes soon.

niteshsinha17 commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,240 @@
+    def create_context(self, request):
+        app_label, model = self.model.split('.')
+        model_label = model.lower()
+        url = reverse(f'admin:{app_label}_{model_label}_{self.name}')
```
@nemesisdesign , @atb00ker and @kkreitmair I found an issue here. It will give an error at runtime if reverse is not found. We can't check reverse before loading of url config. 
I think we should write a try statement here and should remove the item from the menu if reverse is not found.
What do you think?

atb00ker replied:
Hmmm, this can only happen when the app starts for the first time, correct? So let's let the admin see the error so that they can correct their configuration.
It will never be actually visible to the end-user so it shouldn't be a problem as far as I can think...

atb00ker reviewed (commented):
Hi,

I pulled this again and tried, but this is not fixed: https://github.com/openwisp/openwisp-utils/pull/175#pullrequestreview-670415003

This PR was in ready for review tab,

nemesisdesign commented on openwisp_utils/admin_theme/context_processor.py:
```diff
@@ -1,14 +1,24 @@
 from . import settings as app_settings
 
 
-def menu_items(request):
```
This breaks backward compatibility. I tried running this branch on openwisp-controller and I got this exception:

```python
ImportError at /admin/
Module "openwisp_utils.admin_theme.context_processor" does not define a "menu_items" attribute/class
``` 

Please ensure there's an alias, eg: 

```python
# Kept for backward compatibility
# Todo: remove in version 0.9.0
menu_items = menu_groups
```

nemesisdesign commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,240 @@
+
+
+class MenuGroup(BaseMenuItem):
+    '''
```
use double quotes for docstrings for consistency please

nemesisdesign commented on openwisp_utils/admin_theme/menu.py:
```diff
@@ -0,0 +1,240 @@
+            self.label == other.label
+            and self.url == other.url
+            and self.icon == other.icon
+        )
```
nice, what's this for?

niteshsinha17 replied:
I added this to compare items during testing of their order. Today I found a way to do the same in a much simpler way so I am removing this code.

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+^^^^^^^^^^^^^^^^^^^^^^^^
 
-Users will only be able to see menu items for objects they have permission to either view or edit.
+It allow us to register different menu item or group at different position in the ``Main Navigation Menu``.
```
It allows*

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+
+Note: ``register_menu_items`` is deprecated by ``register_menu_group`` and will be removed in the
+future versions. ``register_menu_items`` will be shown at the top of navigation menu and above 
+any register_menu_group items.
```
register_menu_group is missing the code formatting (double slant quotes).

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+future versions. ``register_menu_items`` will be shown at the top of navigation menu and above 
+any register_menu_group items.
+
+``Create a general link``
```
This is not code, therefore it should not be formatted with the double slant quotes around.

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+| ``icon``         | (``str``) CSS class name of the link icon.  (``optional``)   |
++------------------+--------------------------------------------------------------+
+
+``Create a model link``
```
same here

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+| ``icon``         | (``str``) CSS class name of the link icon.  (``optional``)   |
++------------------+--------------------------------------------------------------+
+
+``Create a menu group``
```
same here

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
++------------------+--------------------------------------------------------------+
+| ``items``        | (``dict``) Items to be displayed in the dropdown.            |
+|                  | It can be a dict of ``general link`` or ``model link``       |
+|           .      | with key as their position.                                  |
```
something is wrong here, please look at the generated output: https://github.com/niteshsinha17/openwisp-utils/tree/gsoc/new-navigation-menu-and-ux-improvements#create-a-menu-group
It does not look right. Please after writing docs always look at the generated output and ensure there are no quirks.

niteshsinha17 replied:
Fixed

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
++------------------+--------------------------------------------------------------+
+| ``model``        | (``str``) Model of a app.  (``required``)                    |
++------------------+--------------------------------------------------------------+
+| ``name``         | (``str``) url name. eg. changelist or add.  (``required``)   +
```
are you sure that plus sign at the end is right?

niteshsinha17 replied:
> Hi,
> 
> I pulled this again and tried, but this is not fixed: [#175 (review)](https://github.com/openwisp/openwisp-utils/pull/175#pullrequestreview-670415003)
> 
> This PR was in ready for review tab,

It's working correctly. I fixed it last time. for icons, let's do it later in https://github.com/openwisp/openwisp-utils/issues/206 or https://github.com/openwisp/openwisp-utils/pull/185.

nemesisdesign reviewed (changes requested):
Looks a lot better. I will be testing this in conjunction with https://github.com/openwisp/openwisp-users/pull/254/ soon and will leave another review there as well.

See below for my last request for improvements.

niteshsinha17 commented:
> Hi,
> 
> I pulled this again and tried, but this is not fixed: [#175 (review)](https://github.com/openwisp/openwisp-utils/pull/175#pullrequestreview-670415003)
> 
> This PR was in ready for review tab,

@atb00ker  It's working correctly. I fixed it last time. for icons, let's do it later in https://github.com/openwisp/openwisp-utils/issues/206 or https://github.com/openwisp/openwisp-utils/pull/185.

atb00ker reviewed (approved):
LGTM! 😄 
I think we should move forward, if we face issues later, we will tackle them accordingly.

@niteshsinha17 also, have you tried breaking tests as Fed has asked in the mailing list?

niteshsinha17 commented:
> LGTM! 
> I think we should move forward, if we face issues later, we will tackle them accordingly.
> 
> @niteshsinha17 also, have you tried breaking tests as Fed has asked in the mailing list?

Yes

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+^^^^^^^^^^^^^^^^^^^^^^^^
 
-Users will only be able to see menu items for objects they have permission to either view or edit.
+It allows us to register different menu item or group at different position in the ``Main Navigation Menu``.
```
again, `Main Navigation Menu` is not code, so it shouldn't be flagged as such.

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+            'name': 'changelist',
+            'label': 'MyModel Change List',
+            'icon': 'my-model-list-class',
+            }
```
indentation is odd here

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+.. code-block:: python
+
+    from openwisp_utils.admin_theme.menu import register_menu_group
+
```
You know what would be great, to import the `ugettext_lazy as _` function here and then use it to flag strings as translatable, that way I can avoid having to repeat to everyone to flag strings as translatable when they register menu items :joy: .

nemesisdesign replied:
Thank you for letting me know, please don't do that, because Gettext scans the code to look for strings that have to be translated, if we do it in that way gettext will not be able to add those strings to the .po file (a file which lists all the strings that are translatable and allows translators to translate them to a specific language, see the [django po files for reference](https://github.com/django/django/blob/main/django/contrib/admin/locale/hi/LC_MESSAGES/django.po), in the near future we'll have translations for openwisp too).

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+any ``register_menu_group`` items.
+
+Create a general link
+^^^^^^^^^^^^^^^^^^^^^
```
I think this one and the other new sections below could be made subsections of the section added above, it would result clear that these are different use cases of the function described above.

What's general BTW? Did you mean generic?

Also, I'm not sold on using "Create", what about: "Adding a generic link"?

niteshsinha17 replied:
Yes its generic. I will change it

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+.. code-block:: python
+
+    my_link_config = {'label': 'Link Label', 'url': 'link_url', 'icon':'my-icon'}
+    register_menu_group(position=1, config=my_link_config)
```
What about simplifing this to:

```python
register_menu_group(position=1, config={
     ...
     ...
})
```

I think this way is readable but leaner.
The same suggestion applies to the other code samples.

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+| ``icon``         | (``str``) CSS class name of the link icon.  (``optional``)   |
++------------------+--------------------------------------------------------------+
+
+Create a model link
```
same here as suggested above

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+        'icon': 'my-model-add-class',
+    }
+    register_menu_group(position=1, config=my_model_list_config)
+    register_menu_group(position=2, config=my_model_add_config)
```
same here regarding code samples

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
++------------------+--------------------------------------------------------------+
+
+Create a menu group
+^^^^^^^^^^^^^^^^^^^
```
same here as above

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+    register_menu_group(
+        position=1,
+        config={'label': 'My Label', 'items': my_group_items, 'icon': 'my-icon-class'},
+    )
```
same suggestion about simplifying code samples, I'd show one nested data structure, should make it clear that the data is nested.

nemesisdesign commented on README.rst:
```diff
@@ -454,16 +454,184 @@ template ``context_processors`` in ``settings.py`` as shown below.
+
+    .icon-class-name:{
+        background: url(imageurl);
+    }
```
I don't understand where this CSS should be added, I guess you're implying it can be done using with custom additional CSS files, can you make this more explicit please? And link the specific README section regarding custom CSS if needed.

niteshsinha17 replied:
We are already having a section for it. https://github.com/openwisp/openwisp-utils#supplying-custom-css-and-js-for-the-admin-theme. Should create one more add a subsection in this?
I think we can create a subsection in Main Navigation Menu section

nemesisdesign replied:
I would turn this in a subsection with the title "How to use custom icons in the menu", then explain that it's just needed to supply a custom CSS as indicated in "Supplying custom CSS and JS for the admin theme" (make this a link) which defines the CSS classes with the custom icons.

nemesisdesign reviewed (changes requested):
Spotted more docs issues.
Please triple check your docs.

niteshsinha17 commented:
@nemesisdesign Thanks for your suggestions.  I have followed your suggestions and have done some more improvements from myside.
